### PR TITLE
Add interpreter for InfeedOp

### LINKED ISSUE #1128: Add interpreter for InfeedOp
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### PR DESCRIPTION
**Pull request by ghpvnist**
We have the following constraints in the spec:

```
(I1) `token` is a `token`.
(I2) `infeed_config` is a constant of type `string`.
(C1) `0 < size(results)`.
(C2) `is_empty(result[:-1])` or `is_tensor(type(results[:-1]))`.
(C3) `is_token(type(results[-1]))`.
```

These constraints will be comprehensively covered by the following tests:

```
I1: a) `token` is not a `token`. (Covered by ODS).
I2: a) `infeed_config` is not a constant of type `string`. (Covered by ODS).
C1: a) `size(results) <= 0`.
C2: a) `is_empty(result[:-1]) = false` and `is_tensor(type(results[:-1])) = false`.
C3: a) `type(results[-1]) != token`.
```

If we drop the "Covered by ODS" pieces, this will leave us with the following test cases:

```
C1: a) `size(results) <= 0`.
C2: a) `is_empty(result[:-1]) = false` and `is_tensor(type(results[:-1])) = false`.
C3: a) `type(results[-1]) != token`.
```

closes #1128

### CODE REVIEW

**Comment by burmako** [comment]
Let's iterate some more on the design for infeed/outfeed once I'm back from vacation.

**Review by GleasonK — CHANGES_REQUESTED** [review]

File: stablehlo/reference/Ops.cpp [hunk 1]
```diff
@@ -1215,6 +1219,21 @@ Tensor evalImagOp(const Tensor &operand, ShapedType resultType) {
...
+                                           Region &region, Scope &scope) {
+  if (!process)
+    llvm::report_fatal_error(
+        "infeed is only supported when run via interpreter.run_parallel");
+
+  auto mnemonic = process->infeed();
+  auto results = eval(region.getParentOfType<ModuleOp>()
+                          .lookupSymbol<func::FuncOp>(mnemonic)
+                          .getBody(),
+                      {}, process, &scope);
```
Comment (GleasonK) [inline stablehlo/reference/Ops.cpp:1232]: Is there anything we should verify about these functions? Does a function with multiple return values work? Does each infeed function require that the function returns the same number of results? Will this error gracefully?

**Review by GleasonK — CHANGES_REQUESTED** [review]

File: stablehlo/reference/Ops.cpp (hunk 1 above)
Comment (ghpvnist) [reply stablehlo/reference/Ops.cpp:1389]: Yes, we should add some checks. Here are the verifications I've added:
1. function should return 1 value
2. function return type should be a tensor
3. infeed attribute should be null or non empty.

File: stablehlo/reference/InterpreterOps.cpp [hunk 2]
```diff
@@ -90,6 +92,32 @@ LogicalResult RunParallelOp::verify() {
...
+      if (!func)
+        return emitOptionalError(getLoc(), "Function ", funcName, " not found");
+
+      if (func.getNumResults() != 1)
+        return emitOptionalError(getLoc(), "Function ", funcName,
+                                 " should return 1 tensor but returns ",
+                                 func.getNumResults());
+
+      for (auto type : func.getResultTypes())
+        if (!type.isa<ShapedType>())
```
Comment (GleasonK) [inline stablehlo/reference/InterpreterOps.cpp:114]: Probably don't need a loop if we've verified that there is only one result?

Can use something like:

```
auto resultType = func.getResultTypes()[0];
if (!resultType.isa<ShapedType>()) error(..., resultType);
```

File: stablehlo/reference/InterpreterOps.cpp [hunk 3]
```diff
@@ -90,6 +92,32 @@ LogicalResult RunParallelOp::verify() {
...
 
+  if (getInfeed()) {
+    if (getInfeed()->empty())
+      return emitOptionalError(
+          getLoc(), "infeed attribute is optional or should not be empty");
+
+    for (auto it = getInfeed()->begin(); it != getInfeed()->end(); ++it) {
+      auto funcName = (*it).dyn_cast<FlatSymbolRefAttr>().getAttr();
+      auto func =
+          (*this)->getParentOfType<ModuleOp>().lookupSymbol<func::FuncOp>(
```
Comment (GleasonK) [inline stablehlo/reference/InterpreterOps.cpp:103]: `(*this)` --> `getOperation()` ?

**Review by GleasonK — CHANGES_REQUESTED** [review]

File: stablehlo/reference/ProcessGrid.cpp [hunk 4]
```diff
@@ -114,11 +114,11 @@ void ProcessGrid::ThreadSafeQueue<T>::push(T input) {
 //===----------------------------------------------------------------------===//
 
 ProcessGrid::ProcessGrid(uint32_t numReplicas, uint32_t numPartitions,
-                         std::queue<StringAttr> *infeed)
+                         std::queue<StringAttr> &infeed)
     : numReplicas_(numReplicas), numPartitions_(numPartitions) {
-  while (!infeed->empty()) {
-    auto head = infeed->front();
-    infeed->pop();
+  while (!infeed.empty()) {
```
[the remark this answers is not in the document]
Comment (ghpvnist) [reply stablehlo/reference/ProcessGrid.cpp]: After adding the constructor, things seemed a bit cramped, so I moved the definitions of ThreadSafeMap and ThreadSafeQueue to a separate namespace.

File: stablehlo/reference/InterpreterOps.cpp [hunk 5]
```diff
@@ -92,16 +92,16 @@ LogicalResult RunParallelOp::verify() {
...
+    if (infeed->empty())
       return emitOptionalError(
           getLoc(), "infeed attribute is optional or should not be empty");
 
-    for (auto it = getInfeed()->begin(); it != getInfeed()->end(); ++it) {
-      auto funcName = (*it).dyn_cast<FlatSymbolRefAttr>().getAttr();
-      auto func =
-          (*this)->getParentOfType<ModuleOp>().lookupSymbol<func::FuncOp>(
-              funcName);
+    for (auto flatSymbolRefAttr : infeed->getAsRange<FlatSymbolRefAttr>()) {
```
Comment (GleasonK) [inline stablehlo/reference/InterpreterOps.cpp:100]: Didn't know this method, awesome!

File: stablehlo/reference/ProcessGrid.h [hunk 6]
```diff
@@ -31,31 +31,103 @@ limitations under the License.
 namespace mlir {
 namespace stablehlo {
 
+struct ProcessId;
+class RendezvousResult;
+
+namespace {
```
Comment (GleasonK) [inline stablehlo/reference/ProcessGrid.h:37]: Anonymous namespace in a header file is rarely a good idea. Creates a new namespace per translation unit.

https://stackoverflow.com/questions/357564/uses-for-anonymous-namespaces-in-header-files

I think keeping in ProcessGrid is OK, the alternative would be to add a `detail` namespace, which tells users that these things are implementation details of process grid and should not be relied upon, this is also an OK option. We use `detail` namespace in `Tensor.h`.

**Review by GleasonK — CHANGES_REQUESTED** [review]

File: stablehlo/reference/ProcessGrid.h (hunk 6 above)
Comment (ghpvnist) [reply stablehlo/reference/ProcessGrid.h:37]: I also thought about taking this out of ProcessGrid to something like ThreadSafeUtils.h. These are generic enough to be separated, but this is also exclusively used by ProcessGrid, and I don't anticipate any other uses of these classes. Would be good to add `detail` to be consistent for now though.

File: stablehlo/dialect/TypeInference.cpp [hunk 7]
```diff
@@ -3539,18 +3539,27 @@ LogicalResult verifyDynamicReshapeOp(std::optional<Location> location,
...
                              ValueRange results) {
   auto resultTypes = results.getType();
+  // infeed_c1
   if (resultTypes.empty())
     return emitOptionalError(
         location, "result is expected to be at least of size 1, but got ",
         resultTypes.size());
 
+  // infeed_c2
+  for (size_t i = 0; i < resultTypes.size() - 1; ++i)
```
[the remark this answers is not in the document]
Comment (GleasonK) [reply stablehlo/dialect/TypeInference.cpp:3578]: I believe you can use:

`for (auto resultType : results.drop_back().getTypes())`

and 

`results.back().getType()`

Consistent with the send/receive PR

File: stablehlo/reference/InterpreterOps.td [hunk 8]
```diff
@@ -64,13 +66,15 @@ def Interpreter_RunParallelOp : Op<Interpreter_Dialect, "run_parallel", []> {
     Example:
     ```mlir
     %results:2 = "interpreter.run_parallel"() {
-      programs=[["foo"], ["bar"]]
+      infeed=[%infeed_queue0, @infeed_queue1]
```
Comment (GleasonK) [inline stablehlo/reference/InterpreterOps.td:69]: typo `@infeed_queue0`

**Review by GleasonK — APPROVED** [review]
Last bit of feedback then LGTM

File: stablehlo/reference/ProcessGrid.h (hunk 6 above)
Comment (GleasonK) [reply stablehlo/reference/ProcessGrid.h]: Detail is fine. Prefer to not make things overly generic/reusable until there is a need for it

File: stablehlo/reference/InterpreterOps.td (hunk 8 above)
Comment (ghpvnist) [reply stablehlo/reference/InterpreterOps.td:69]: Thanks!

**Pull request merged**